src/nvim/diff.c

```diff
@@ -1553,73 +1617,20 @@ static void diff_read(int idx_orig, int idx_new, diffio_T *dio)
     }
   }
 
-  diffhunk_T hunk;
-
   for (;;) {
+    diffhunk_T hunk;
+    bool eof = false;
     if (dio->dio_internal) {
-      if (line_idx >= dout->dout_ga.ga_len) {
-        break;      // did last line
+      eof = line_idx >= dout->dout_ga.ga_len;
+      if (!eof) {
+        hunk = ((diffhunk_T *)dout->dout_ga.ga_data)[line_idx++];
       }
-      hunk = ((diffhunk_T *)dout->dout_ga.ga_data)[line_idx++];
     } else {
-      char *line;
-      char linebuf[LBUFLEN];  // only need to hold the diff line
-      if (fd == NULL) {
-        if (line_idx >= dout->dout_ga.ga_len) {
-          break;      // did last line
-        }
-        line = ((char **)dout->dout_ga.ga_data)[line_idx++];
-      } else {
-        if (vim_fgets((char_u *)linebuf, LBUFLEN, fd)) {
-          break;      // end of file
-        }
-        line = linebuf;
-      }
-
-      if (diffstyle == DIFF_NONE) {
-        // Determine diff style.
-        // ed like diff looks like this:
-        // {first}[,{last}]c{first}[,{last}]
-        // {first}a{first}[,{last}]
-        // {first}[,{last}]d{first}
-        //
-        // unified diff looks like this:
-        // --- file1       2018-03-20 13:23:35.783153140 +0100
-        // +++ file2       2018-03-20 13:23:41.183156066 +0100
-        // @@ -1,3 +1,5 @@
-        if (isdigit(*line)) {
-          diffstyle = DIFF_ED;
-        } else if ((STRNCMP(line, "@@ ", 3) == 0)) {
-          diffstyle = DIFF_UNIFIED;
-        } else if ((STRNCMP(line, "--- ", 4) == 0)  // -V501
-                   && (vim_fgets((char_u *)linebuf, LBUFLEN, fd) == 0)  // -V501
-                   && (STRNCMP(line, "+++ ", 4) == 0)
-                   && (vim_fgets((char_u *)linebuf, LBUFLEN, fd) == 0)  // -V501
-                   && (STRNCMP(line, "@@ ", 3) == 0)) {
-          diffstyle = DIFF_UNIFIED;
-        } else {
-          // Format not recognized yet, skip over this line.  Cygwin diff
-          // may put a warning at the start of the file.
-          continue;
-        }
-      }
+      hunk = extract_hunk(fd, &diffstyle, &eof);
```

Comment:
How about passing `hunk` as a pointer and using `eof` as return value? That seems less costly.

Reply:
The two `break` statements in function can also then be changed to `return` statements, thus avoiding returning an uninitialized value.

Reply:
I've done the first suggestion. Not sure what you fully mean by the second. There was only 1 break statement on eof.